core/blockchain.go

```diff
@@ -1763,6 +1786,9 @@ func (bc *BlockChain) insertChain(chain types.Blocks, setHead bool) (int, error)
 		if err := bc.validator.ValidateState(block, statedb, receipts, usedGas); err != nil {
 			bc.reportBlock(block, receipts, err)
 			followupInterrupt.Store(true)
+			if bc.logger != nil {
+				bc.logger.OnBlockValidationError(block, err)
+			}
```

Comment:
This must also be called just after line 1780 (so a bit earlier in this file, sorry couldn't comment directly on it).

Indeed, `bc.processor.Process` will have done `OnBlockStart` so it's the good flow to also report the  `OnBlockValidationError` there.

Reply:
I see now why it's not there, within `Process` call we perform `OnBlockEnd` with the error so the flow don't require the `OnBlockValidationError` (which re-reading make sense to apply only to validation step).

But it remains a flow problem of the logger, you cannot know if the block has been written or not. You cannot rely on `OnBlockEnd` because even if `err` there is `nil`, an `OnBlockValidationError` could happen. And `OnBlockValidationError` cannot be used because not call in all paths.

We are missing an `OnBlockWritten` that would let the logger know that the block can be flushed correctly. I have added it in my code after `writeBlockWithState`.

Reply:
I understand your concern. I would have liked to keep `OnBlockStart` and `OnBlockEnd` contained within StateProcessor but seems that's not going to work. So instead of adding another method I will simply move `OnBlockEnd` further to catch those errors and drop `OnBlockValidationError`.

Reply:
Pushed. Do you think this version's ok? BTW I assumed TD is not useful when block importing err'd.

Reply:
I'll update my current work on top of your branch tomorrow and report back.

Reply:
All good about this, resolved from my point of view.